<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>DAseq Tutorial â€¢ DAseq</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="DAseq Tutorial">
<meta property="og:description" content="DAseq">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DAseq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/tutorial.html">DAseq Tutorial</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>DAseq Tutorial</h1>
            
      
      
      <div class="hidden name"><code>tutorial.Rmd</code></div>

    </div>

    
    
<div id="da-seq-tutorial" class="section level2">
<h2 class="hasAnchor">
<a href="#da-seq-tutorial" class="anchor"></a>DA-seq tutorial</h2>
<div id="overview" class="section level3">
<h3 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h3>
<p>This tutorial shows basic steps to run DAseq on scRNA-seq datasets.</p>
<p>Data used in this tutorial is from <a href="https://www.sciencedirect.com/science/article/pii/S0092867418313941">Sade-Feldman, Moshe, et al. (Cell. 2018)</a>. The goal is to compare immune cells from melanoma patients that respond or not respond to immune checkpoint therapy: responders (R) and non-responders (NR), and identify cell subpopulations with differential abundance (DA) between them.</p>
<p>With DAseq, five DA cell subpopulations are detected and characterized.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">DAseq</span>)</pre></body></html></div>
</div>
<div id="set-some-essential-python-parameter" class="section level3">
<h3 class="hasAnchor">
<a href="#set-some-essential-python-parameter" class="anchor"></a>Set some essential Python parameter</h3>
<p>Set Python to use, put the PATH to your Python binary here</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="no">python2use</span> <span class="kw">&lt;-</span> <span class="st">"/path/to/your/python"</span></pre></body></html></div>
<p>Set GPU to use, put the GPU number here</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">GPU</span> <span class="kw">&lt;-</span> <span class="fl">5</span></pre></body></html></div>
</div>
<div id="get-label-information" class="section level3">
<h3 class="hasAnchor">
<a href="#get-label-information" class="anchor"></a>Get label information</h3>
<p>Here, we will get the sample label names for two biological conditions: responders (R) and non-responders (NR) in the data, respectively. In total, there are 48 samples, 17 of which are responders (R) and 31 of which are non-responders (NR). This information is in the dataframe <code>X.label.info</code> attached to the DAseq package: the label column shows the sample label name, and the condition column shows the condition (R or NR).</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">X.label.info</span>)</pre></body></html></div>
<pre><code>##       label condition
## 1    Pre_P1         R
## 2   Post_P1         R
## 3   Post_P4         R
## 4 Post_P5_2         R
## 5    Pre_P7         R
## 6   Post_P7         R</code></pre>
<p>Now we get the sample label names for R and NR.</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">labels_res</span> <span class="kw">&lt;-</span> <span class="no">X.label.info</span>[<span class="no">X.label.info</span>$<span class="no">condition</span> <span class="kw">==</span> <span class="st">"R"</span>, <span class="st">"label"</span>]
<span class="no">labels_nonres</span> <span class="kw">&lt;-</span> <span class="no">X.label.info</span>[<span class="no">X.label.info</span>$<span class="no">condition</span> <span class="kw">==</span> <span class="st">"NR"</span>, <span class="st">"label"</span>]</pre></body></html></div>
</div>
<div id="get-da-cells" class="section level3">
<h3 class="hasAnchor">
<a href="#get-da-cells" class="anchor"></a>Get DA cells</h3>
<p>This step incorporates the first two steps of DAseq algorithm. The input of this step includes:</p>
<ul>
<li>
<code>X</code>, the scRNA-seq data after dimension reduction (PCA), this is included in the DAseq package as <code>X.melanoma</code>
</li>
<li>
<code>cell.labels</code>, sample label for every cell in the data, this is included in the DAseq package as <code>X.label.melanoma</code>
</li>
<li>
<code>labels.1</code>, sample label names that correspond to one biological condition (R), <code>R_labels</code>
</li>
<li>
<code>labels.2</code>, sample label names that correspond to the other biological condition (NR), <code>NR_labels</code>
</li>
<li>
<code>k.vector</code>, values of <em>k</em> to use for the calculation of score vector with <em>kNN</em>, here we will use <em>k</em> from 50 to 500</li>
<li>
<code>plot.embedding</code>, 2D embedding of the data to visualize the results (t-SNE), this is included in the DAseq package as <code>X.2d.melanoma</code>
</li>
</ul>
<p>Runtime for this step with <code>n.runs = 1</code> is less than a minute on a standard laptop, and it will scale linearly with <code>n.runs</code>.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="no">da_cells</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/getDAcells.html">getDAcells</a></span>(
  <span class="kw">X</span> <span class="kw">=</span> <span class="no">X.melanoma</span>,
  <span class="kw">cell.labels</span> <span class="kw">=</span> <span class="no">X.label.melanoma</span>,
  <span class="kw">labels.1</span> <span class="kw">=</span> <span class="no">labels_res</span>,
  <span class="kw">labels.2</span> <span class="kw">=</span> <span class="no">labels_nonres</span>,
  <span class="kw">k.vector</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="fl">50</span>, <span class="fl">500</span>, <span class="fl">50</span>),
  <span class="kw">plot.embedding</span> <span class="kw">=</span> <span class="no">X.2d.melanoma</span>
)</pre></body></html></div>
<p>The output of this step is a list of results:</p>
<ul>
<li>
<code>da.ratio</code>, score vector for each cell</li>
<li>
<code>da.pred</code>, DA measure for each cell</li>
<li>
<code>da.up</code>, index of DA cells more abundant in <em>labels.2</em> (non-responders)</li>
<li>
<code>da.down</code>, index of DA cells more abundant in <em>labels.1</em> (responders)</li>
</ul>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span>(<span class="no">da_cells</span>[<span class="fl">1</span>:<span class="fl">2</span>])</pre></body></html></div>
<pre><code>## List of 2
##  $ da.ratio: num [1:16291, 1:10] -0.51756 0.29553 0.00342 -0.20471 -0.38711 ...
##   ..- attr(*, "dimnames")=List of 2
##   .. ..$ : NULL
##   .. ..$ : chr [1:10] "50" "100" "150" "200" ...
##  $ da.pred : num [1:16291] -0.5088 0.3308 0.0449 -0.0564 -0.5612 ...</code></pre>
<p>The prediction values are overlayed on the 2D embedding in the <code>pred.plot</code> slot of the output.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="no">da_cells</span>$<span class="no">pred.plot</span></pre></body></html></div>
<p><img src="tutorial_files/figure-html/unnamed-chunk-7-1.png" width="480"></p>
<p>DAseq runs a random permutation on the labels to generate the null distribution for DA measure. We can check the permutation results in the <code>rand.plot</code> slot.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">da_cells</span>$<span class="no">rand.plot</span></pre></body></html></div>
<p><img src="tutorial_files/figure-html/unnamed-chunk-8-1.png" width="432"></p>
<p>By default, cells with DA measure values larger than the maximum or small than the minumum than permutated DA measure are selected as DA cells. The threshold can be changed by running <code><a href="../reference/updateDAcells.html">updateDAcells()</a></code> to look at the most salient DA cells. In this data, we use <strong>0.8</strong> (for non-responder) and <strong>-0.8</strong> (for responder).</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="no">da_cells</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/updateDAcells.html">updateDAcells</a></span>(
  <span class="kw">X</span> <span class="kw">=</span> <span class="no">da_cells</span>, <span class="kw">pred.thres</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(-<span class="fl">0.8</span>,<span class="fl">0.8</span>),
  <span class="kw">plot.embedding</span> <span class="kw">=</span> <span class="no">X.2d.melanoma</span>
)</pre></body></html></div>
<p>Selected DA cells are highlighted in the 2D embedding in the <code>da.cells.plot</code> slot. Cells in red are in the top (these cells are more abundant in non-responder samples), blue in the bottom (these cells are more abundant in responder samples).</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="no">da_cells</span>$<span class="no">da.cells.plot</span></pre></body></html></div>
<p><img src="tutorial_files/figure-html/unnamed-chunk-10-1.png" width="432"></p>
</div>
<div id="get-da-regions" class="section level3">
<h3 class="hasAnchor">
<a href="#get-da-regions" class="anchor"></a>Get DA regions</h3>
<p>In this step, selected DA cells will be clustered into several coherent regions, which represent the DA cell subpopulations. Several input parameters are identical to that of <code><a href="../reference/getDAcells.html">getDAcells()</a></code>: <code>X, cell.labels, labels.1, labels.2, plot.embedding</code>.</p>
<p>Other input includes:</p>
<ul>
<li>
<code>da.cells</code>: the output of <code><a href="../reference/getDAcells.html">getDAcells()</a></code>
</li>
<li>
<code>resolution</code>: clustering parameter, default 0.05, use a larger value to group DA cells into more regions</li>
<li>
<code>min.cell</code>: minimum number of cells, regions with fewer cells will be removed, default is to use the minumum <em>k</em> value in the <code>k.vector</code>
</li>
</ul>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="no">da_regions</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/getDAregion.html">getDAregion</a></span>(
  <span class="kw">X</span> <span class="kw">=</span> <span class="no">X.melanoma</span>,
  <span class="kw">da.cells</span> <span class="kw">=</span> <span class="no">da_cells</span>,
  <span class="kw">cell.labels</span> <span class="kw">=</span> <span class="no">X.label.melanoma</span>,
  <span class="kw">labels.1</span> <span class="kw">=</span> <span class="no">labels_res</span>,
  <span class="kw">labels.2</span> <span class="kw">=</span> <span class="no">labels_nonres</span>,
  <span class="kw">resolution</span> <span class="kw">=</span> <span class="fl">0.01</span>,
  <span class="kw">plot.embedding</span> <span class="kw">=</span> <span class="no">X.2d.melanoma</span>,
)</pre></body></html></div>
<p>The output is a list of results:</p>
<ul>
<li>
<code>da.region.label</code>, DA subpopulation labels for all the cells, 0 means non-DA cells</li>
<li>
<code>DA.stat</code>, statistics of each DA subpopulation</li>
</ul>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span>(<span class="no">da_regions</span>[<span class="fl">1</span>:<span class="fl">2</span>])</pre></body></html></div>
<pre><code>## List of 2
##  $ da.region.label: num [1:16291] 0 0 0 0 0 0 0 0 0 0 ...
##  $ DA.stat        : num [1:5, 1:3] 0.973 0.965 -0.933 -0.976 -0.904 ...
##   ..- attr(*, "dimnames")=List of 2
##   .. ..$ : NULL
##   .. ..$ : chr [1:3] "DA.score" "pval.wilcoxon" "pval.ttest"</code></pre>
<p>Clustering result is shown in the plot: <code>da.region.plot</code> slot.</p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="no">da_regions</span>$<span class="no">da.region.plot</span></pre></body></html></div>
<p><img src="tutorial_files/figure-html/unnamed-chunk-13-1.png" width="480"></p>
</div>
<div id="get-markers-for-each-da-subpopulation-with-stg" class="section level3">
<h3 class="hasAnchor">
<a href="#get-markers-for-each-da-subpopulation-with-stg" class="anchor"></a>Get markers for each DA subpopulation with STG</h3>
<p>The final step of DAseq is to characterize each DA subpopulation by detecting genes that seprate the DA subpopulation from the rest of the cells through STG (stochastic gates).</p>
<p>The gene expression data is NOT included in the DAseq package and needs to be downloaded. This code will download the data to the current working directory, and load this data into R.</p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span>(
  <span class="st">"ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE120nnn/GSE120575/suppl/GSE120575_Sade_Feldman_melanoma_single_cells_TPM_GEO.txt.gz"</span>,
  <span class="st">"./GSE120575_Sade_Feldman_melanoma_single_cells_TPM_GEO.txt.gz"</span>
)
<span class="no">X.data.melanoma</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(
  <span class="st">"./GSE120575_Sade_Feldman_melanoma_single_cells_TPM_GEO.txt.gz"</span>,
  <span class="kw">sep</span> <span class="kw">=</span> <span class="st">"\t"</span>, <span class="kw">header</span> <span class="kw">=</span> <span class="no">F</span>, <span class="kw">row.names</span> <span class="kw">=</span> <span class="fl">1</span>, <span class="kw">stringsAsFactors</span> <span class="kw">=</span> <span class="no">F</span>, <span class="kw">skip</span> <span class="kw">=</span> <span class="fl">2</span>
)
<span class="no">X.data.melanoma</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="no">X.data.melanoma</span>[,-<span class="fl">16292</span>])</pre></body></html></div>
<p>Then, we will use STG to identify markers for each DA subpopulation. The input parameters for STG are the normalized gene expression matrix, and the output from the function <code><a href="../reference/getDAregion.html">getDAregion()</a></code>.</p>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="no">STG_markers</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/STGmarkerFinder.html">STGmarkerFinder</a></span>(
  <span class="kw">X</span> <span class="kw">=</span> <span class="no">X.data.melanoma</span>,
  <span class="kw">da.regions</span> <span class="kw">=</span> <span class="no">da_regions</span>,
  <span class="kw">lambda</span> <span class="kw">=</span> <span class="fl">1.5</span>, <span class="kw">n.runs</span> <span class="kw">=</span> <span class="fl">5</span>, <span class="kw">return.model</span> <span class="kw">=</span> <span class="no">T</span>,
  <span class="kw">python.use</span> <span class="kw">=</span> <span class="no">python2use</span>, <span class="kw">GPU</span> <span class="kw">=</span> <span class="no">GPU</span>
)</pre></body></html></div>
<p>The output is a list of results:</p>
<ul>
<li>
<code>da.markers</code>, a list of data frames, each data frame contains markers for each DA subpopulation</li>
<li>
<code>accuracy</code>, accuracy of STG for each DA subpopulation</li>
<li>
<code>model</code>, the actual model from STG for each DA subpopulation, please refer to the documentation of <code><a href="../reference/STGmarkerFinder.html">STGmarkerFinder()</a></code>
</li>
</ul>
<p>Top markers for DA region 5 are shown:</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">STG_markers</span>$<span class="no">da.markers</span><span class="kw">[[</span><span class="st">"5"</span>]])</pre></body></html></div>
<pre><code>##        gene avg_logFC       p_value
## TCF7   TCF7 1.8865021 4.261047e-122
## RPL13 RPL13 0.1765772 2.421585e-113
## RPL3   RPL3 0.2037486 1.516596e-102
## LEF1   LEF1 2.3623672 6.638429e-102
## RPS6   RPS6 0.2002730 2.013261e-100
## RPS3A RPS3A 0.2040410  2.452806e-98</code></pre>
<p>In the <code>model</code> slot, predictions from STG are available. These predictions are linear combinations of a limited set of genes that best separate the DA subpopulation from the rest of the cells.</p>
<p>For example, here we will plot the prediction value from STG for DA subpopulation 5: the <code>pred</code> slot of the 5th list in the <code>model</code> slot of the output.</p>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="fu"><a href="../reference/plotCellScore.html">plotCellScore</a></span>(
  <span class="kw">X</span> <span class="kw">=</span> <span class="no">X.2d.melanoma</span>, <span class="kw">score</span> <span class="kw">=</span> <span class="no">STG_markers</span>$<span class="no">model</span><span class="kw">[[</span><span class="st">"5"</span>]]$<span class="no">pred</span>
)</pre></body></html></div>
<p><img src="tutorial_files/figure-html/unnamed-chunk-17-1.png" width="480"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Jun Zhao.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
